### PR TITLE
Chapter1/homework/7

### PR DESCRIPTION
**Pull request by mihailfox**
I implemented all methods and I added demo for each of the four methods in Main()

### CODE REVIEW

File: Src/BootCamp.Chapter/Program.cs [hunk 1]
```diff
@@ -1,14 +1,35 @@
-﻿namespace BootCamp.Chapter
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  4
         

Clones added
============
- Src/BootCamp.Chapter/BalanceStats.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157&bid=16419471)

File: Src/BootCamp.Chapter/Program.cs [hunk 2]
```diff
@@ -1,14 +1,35 @@
-﻿namespace BootCamp.Chapter
+﻿using System;
+using System.Collections.Generic;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 3]
```diff
@@ -1,37 +1,422 @@
...
 
         /// <summary>
         /// Return name and loss of a person with a biggest loss (balance change negative).
         /// </summary>
         public static string FindPersonWithBiggestLoss(string[] peopleAndBalances)
         {
-            return "";
+            if (!ArrayIsValid(peopleAndBalances)) return invalidMessage;
+
+            //const string peopleMessage = " lost the most money. ";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Remove this commented out code.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

File: Src/BootCamp.Chapter/Program.cs [hunk 4]
```diff
@@ -1,14 +1,35 @@
...
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 5]
```diff
@@ -1,37 +1,422 @@
...
+                return resultMessage.ToString();
+            }
+        }
+
+        /// <summary>
+        /// Return name and current money of the most poor person.
+        /// </summary>
+        public static string FindMostPoorPerson(string[] peopleAndBalances)
+        {
+            if (!ArrayIsValid(peopleAndBalances)) return invalidMessage;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

File: Src/BootCamp.Chapter/Program.cs [hunk 6]
```diff
@@ -1,14 +1,35 @@
-﻿namespace BootCamp.Chapter
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 7]
```diff
@@ -1,37 +1,422 @@
...
+using System.Text;
+
+namespace BootCamp.Chapter
 {
     public static class BalanceStats
     {
+        private const string currencySymbol = "¤";
+        private const string invalidMessage = "N/A.";
+        private const string messageEnd = ".";
+        private static readonly NumberFormatInfo numberFormatInfo = new NumberFormatInfo() { NumberDecimalSeparator = messageEnd };
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Remove these redundant parentheses.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 8]
```diff
@@ -1,37 +1,422 @@
...
+        /// </summary>
+        private static bool InputStringIsValid(string inputString)
+        {
+            return !string.IsNullOrEmpty(inputString) && !string.IsNullOrWhiteSpace(inputString);
+        }
+
+        /// <summary>
+        /// Finds lowest balance for a single person.
+        /// </summary>
+        private static decimal LowestBalanceForSinglePerson(string personAndBalance)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Remove the unused private method 'LowestBalanceForSinglePerson'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

Comment (mihailfox) [reply Src/BootCamp.Chapter/BalanceStats.cs]: Fixed in [2db6c28](https://github.com/csinn/CSharp-From-Zero-To-Hero/pull/161/commits/2db6c282219933cc661146c542a210b0f73f44cf)

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 7 above)
Comment (mihailfox) [reply Src/BootCamp.Chapter/BalanceStats.cs]: Fixed in [2db6c28](https://github.com/csinn/CSharp-From-Zero-To-Hero/pull/161/commits/2db6c282219933cc661146c542a210b0f73f44cf)

File: Src/BootCamp.Chapter/Program.cs (hunk 6 above)
Comment (mihailfox) [reply Src/BootCamp.Chapter/Program.cs]: Fixed in [2db6c28](https://github.com/csinn/CSharp-From-Zero-To-Hero/pull/161/commits/2db6c282219933cc661146c542a210b0f73f44cf)

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 5 above)
Comment (mihailfox) [reply Src/BootCamp.Chapter/BalanceStats.cs]: Fixed in [2db6c28](https://github.com/csinn/CSharp-From-Zero-To-Hero/pull/161/commits/2db6c282219933cc661146c542a210b0f73f44cf)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  4
         

Clones added
============
- Src/BootCamp.Chapter/BalanceStats.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157&bid=16419471)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 9]
```diff
@@ -1,37 +1,409 @@
...
+        private const string messageEnd = ".";
+        private static readonly NumberFormatInfo numberFormatInfo = new NumberFormatInfo { NumberDecimalSeparator = messageEnd };
+
         /// <summary>
         /// Return name and balance(current) of person who had the biggest historic balance.
         /// </summary>
         public static string FindHighestBalanceEver(string[] peopleAndBalances)
         {
-            return "";
+            if (!ArrayIsValid(peopleAndBalances)) return invalidMessage;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  4
         

Clones added
============
- Src/BootCamp.Chapter/BalanceStats.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157&bid=16419471)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 10]
```diff
@@ -1,37 +1,412 @@
...
+            return resultMessage.ToString();
         }
 
         /// <summary>
         /// Return name and current money of the richest person.
         /// </summary>
         public static string FindRichestPerson(string[] peopleAndBalances)
         {
-            return "";
+            if (!ArrayIsValid(peopleAndBalances)) return invalidMessage;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  4
         

Clones added
============
- Src/BootCamp.Chapter/BalanceStats.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157&bid=16419471)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157&bid=16419471)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157&bid=16419471)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 11]
```diff
@@ -1,37 +1,421 @@
-﻿namespace BootCamp.Chapter
+﻿using System.Globalization;
+using System.Text;
+
+namespace BootCamp.Chapter
 {
     public static class BalanceStats
     {
+        private const string currencySymbol = "¤";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Well done using constants.

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 12]
```diff
@@ -1,37 +1,421 @@
...
+            }
+
+            var highestBalanceIndex = FindDecimalArrayMax(highestBalanceList);
+            var highestBalance = highestBalanceList[highestBalanceIndex];
+            var personWithHighestBalance = peopleList[highestBalanceIndex];
+
+            if (ArrayElementsAreEqual(highestBalanceList))
+            {
+                resultMessage
+                    .Append(FormatStringAndCommas(peopleList))
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Neat formatting.

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 13]
```diff
@@ -1,37 +1,421 @@
...
+            {
+                mostPoorBalanceList[i] = CurrentBalanceForSinglePerson(peopleAndBalances[i]);
+                mostPoorPeopleList[i] = ReturnNameForSinglePerson(peopleAndBalances[i]);
+            }
+
+            var mostPoorPersonsIndex = FindDecimalArrayMin(mostPoorBalanceList);
+            var mostPoortPerson = mostPoorPeopleList[mostPoorPersonsIndex];
+            var mostPoorPersonsMoney = mostPoorBalanceList[mostPoorPersonsIndex];
+
+            if (mostPoorBalanceList.Length > 2 && ArrayElementsAreEqual(mostPoorBalanceList))
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Instead of magic numbers use named constants

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 11 above)
Comment (mihailfox) [reply Src/BootCamp.Chapter/BalanceStats.cs]: Cheers!

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 12 above)
Comment (mihailfox) [reply Src/BootCamp.Chapter/BalanceStats.cs]: Thank you :)

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 13 above)
Comment (mihailfox) [reply Src/BootCamp.Chapter/BalanceStats.cs]: I fixed it in commit [529d3c2](https://github.com/csinn/CSharp-From-Zero-To-Hero/pull/161/commits/529d3c25f3eb08ddb62057f8abd449b799f368f7).
Thank you!

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4970157&bid=16419471)

**Pull request closed without merging**